### PR TITLE
dan/per-10153-change-local-facts-uploader-to-use-update-callbacks

### PR DESCRIPTION
**Pull request by danyi1212**

### CODE REVIEW

File: horizon/facts/update_subscriber.py [hunk 1]
```diff
@@ -1,58 +1,45 @@
...
+        reporter.report_update_results = self.decorator(reporter.report_update_results)
 
-    async def _on_message(self, topic: str = "", data: dict | None = None) -> None:
-        """
-        Callback for incoming messages from the PubSub client.
-        """
-        if data is None:
-            logger.debug(f"Received message on topic {topic!r} without data")
-            return
+    def decorator(self, func):
```
Comment (omer9564) [inline horizon/facts/update_subscriber.py:25]: give this function better naming

File: horizon/facts/update_subscriber.py [hunk 2]
```diff
@@ -1,58 +1,45 @@
...
+        @wraps(func)
+        async def wrapper(report: DataUpdateReport, *args, **kwargs):
```
Comment (omer9564) [inline horizon/facts/update_subscriber.py:27]: the original function has "self" as first argument, from what I remember report would be of type CallbacksReporter with the current code ( as it is the first argument passed to wrapper )

File: horizon/facts/update_subscriber.py [hunk 3]
```diff
@@ -1,58 +1,45 @@
...
+            await self._resolve_listeners(report.update_id)
```
Comment (omer9564) [inline horizon/facts/update_subscriber.py:28]: report.update_id can be None, make sure to handle this

File: horizon/facts/update_subscriber.py [hunk 4]
```diff
@@ -1,58 +1,45 @@
...
+            return await func(report, *args, **kwargs)
```
Comment (omer9564) [inline horizon/facts/update_subscriber.py:29]: you can do both of them concurrently, am I right ?

**Review by omer9564 — CHANGES_REQUESTED** [review]
Left few comments, good approach in general.
Besides of that, we added a sleep to mitigate the gap between the pub/sub ack and actual engine ack.
Reduce the default for this configurable sleep

File: horizon/facts/update_subscriber.py (hunk 2 above)
Comment (danyi1212) [reply horizon/facts/update_subscriber.py:27]: It works like this when the report is at first argument. I think python makes it a bound function when the class initiates, so func at this point is already a bound method so the arguments does not include self anymore

File: horizon/facts/update_subscriber.py (hunk 4 above)
Comment (danyi1212) [reply horizon/facts/update_subscriber.py:29]: Well technically the resolve listeners does not need to be async, it's just from PyCharm refactoring

File: horizon/facts/update_subscriber.py (hunk 1 above)
Comment (omer9564) [reply horizon/facts/update_subscriber.py:25]: WDYT on `_report_callback_decorator` ( would also make it `@staticmethod` )

File: horizon/facts/update_subscriber.py (hunk 2 above)
Comment (omer9564) [reply horizon/facts/update_subscriber.py:27]: Weird, if it works it is fine by me, to be honest I expected it to fail 😅

File: horizon/facts/update_subscriber.py [hunk 5]
```diff
@@ -25,12 +25,15 @@ def _inject_subscriber(self):
     def decorator(self, func):
         @wraps(func)
         async def wrapper(report: DataUpdateReport, *args, **kwargs):
-            await self._resolve_listeners(report.update_id)
+            if report.update_id is not None:
+                self._resolve_listeners(report.update_id)
+            else:
+                logger.debug("Received update report without update ID")
```
Comment (omer9564) [inline horizon/facts/update_subscriber.py:31]: Not sure if I would call this before or after the regular callbacks 🤔

**Review by omer9564 — APPROVED** [review]
LGTM, left some semantic non-blocking comments

**Comment by omer9564** [comment]
Don't forget to reduce the default sleep time ( I would personally put 0 because it is no longer required )

File: horizon/facts/update_subscriber.py (hunk 1 above)
Comment (danyi1212) [reply horizon/facts/update_subscriber.py:25]: This cannot be static as it uses `self`

File: horizon/facts/update_subscriber.py (hunk 5 above)
Comment (danyi1212) [reply horizon/facts/update_subscriber.py]: For performance sake, before is better IMO

**Pull request merged**